File: python/813. Largest Sum of Averages.py
class Solution(object):
    def largestSumOfAverages(self, A, K):
        """
        :type A: List[int]
        :type K: int
        :rtype: float
        """
        N = len(A)
        dp = [[0]*(K+1) for _ in range(N+1)]
        s = [0]*(N+1)
        for i in range(1, N+1):
            s[i] = s[i-1] + A[i-1]
            dp[i][1] = float(s[i])/i

        for i in range(1, N+1):
            for k in range(2, K+1):
                for j in range(i):
                    dp[i][k] = max(dp[i][k], dp[j][k-1] + (s[i]-s[j])/float(i-j))
        return dp[N][K]
        

# dp[i][k]表示前i个数构成k个子数组时的最大平均值, 那么对于所有 0 <= j <= i-1
# dp[i][k] = Math.max(dp[i][k], dp[j][k-1] + (A[j+1] + ... + A[i+1]) / (i-j))

File: go/src/leetcode/813/813. Largest Sum of Averages.go
/*
*
@author ZhengHao Lou
Date    2022/11/28
*/
package main

import "fmt"

/*
*
https://leetcode.cn/problems/largest-sum-of-averages/
*/
func largestSumOfAverages(nums []int, K int) float64 {
	n := len(nums)
	f := make([][]float64, n+1)
	for i := range f {
		f[i] = make([]float64, K+1)
	}
	s := make([]int, n+1)
	for i := range nums {
		s[i+1] = s[i] + nums[i]
		f[i+1][1] = float64(s[i+1]) / float64(i+1)
	}
	for i := 1; i <= n; i++ {
		for k := 2; k <= K; k++ {
			for j := 0; j < i; j++ {
				f[i][k] = max(f[i][k], f[j][k-1]+float64(s[i]-s[j])/float64(i-j))
			}
		}
	}

	fmt.Println(f[n][K])
	return f[n][K]
}

func max(a, b float64) float64 {
	if a > b {
		return a
	}
	return b
}

func main() {
	largestSumOfAverages([]int{9, 1, 2, 3, 9}, 3)
}
